Update SDK docs for latest versions (#1808)
astronomer/docs: learn/astro-python-sdk-etl.md

These functions make your DAGs easier to write and read with less code. In this guide, you’ll learn about how to install the Python SDK and how to use it in practice. The Astro SDK should feel more similar to writing a traditional Python script than writing a DAG in Airflow.

:::info

> This guide is based on the latest version of the Astro SDK. You might need to modify the configuration steps or code if you're using an earlier version of the SDK. See the [Astro Python SDK Changelog](https://astro-sdk-python.readthedocs.io/en/1.4/CHANGELOG.html#) for a complete list of changes in each version.

Reviewer: ```suggestion
The latest version of the Astro SDK was used to create this guide. If you're using an earlier version of the Astro SDK, you might need to modify the configuration steps or code. See the [Astro Python SDK Changelog](https://astro-sdk-python.readthedocs.io/en/1.4/CHANGELOG.html#) for a complete list of changes in each version.
```

> This guide is based on the latest version of the Astro SDK. If you're using an earlier version of the Astro SDK, you might need to modify the configuration steps or code. See the [Astro Python SDK Changelog](https://astro-sdk-python.readthedocs.io/en/1.4/CHANGELOG.html#) for a complete list of changes in each version.

Reviewer 2: ```suggestion
This guide is based on the latest version of the Astro SDK. If you're using an earlier version of the Astro SDK, you might need to modify the configuration steps or code. See the [Astro Python SDK Changelog](https://astro-sdk-python.readthedocs.io/en/stable/CHANGELOG.html) for a complete list of changes in each version.
```

Reviewer 2: @jwitz I think a stable release URL is better than a version-specific URL.

Reviewer 3: @utkarsharma2 I just noticed that when we use stable, Bug Fixes appear on the ReadTheDocs menu by the left, could you look into this
![Screenshot from 2023-01-31 11-22-05](https://user-images.githubusercontent.com/272048/215746974-5c19abcf-8f1a-4c88-a4ce-613027ed7bff.png)
![Screenshot from 2023-01-31 11-22-13](https://user-images.githubusercontent.com/272048/215746984-50ff61f2-5ec9-49f8-883c-32eeba7cea82.png)

:::

## Assumed knowledge

To get the most out of this guide, you should have an understanding of Airflow decorators. See [Introduction to Airflow Decorators guide](airflow-decorators.md).

## Python SDK functions

The Astro Python SDK makes implementing ELT use cases easier by allowing you to seamlessly transition between Python and SQL for each step in your process. Details like creating dataframes, storing intermediate results, passing context and data between tasks, and creating task dependencies are all managed automatically.

More specifically, the Astro Python SDK includes several functions that are helpful when implementing an ETL framework:

> - `load_file`: Loads a given file into a SQL table. The file should be in CSV, JSON, or parquet files stored in Amazon S3, Google Drive, or GCS.

Reviewer: ```suggestion
- `load_file`: Loads a given file into a SQL table. The file format should be CSV, JSON, or Parquet and the files should be stored in Amazon S3, Google Drive, or Google Cloud Storage (GCS).
```
I'm assuming Azure isn't an option currently.

> - `load_file`: Loads a given file into a SQL table. The file format should be CSV, JSON, or Parquet and the files should be stored in Amazon S3, Google Drive, or Google Cloud Storage (GCS).

Reviewer 2: We also support NDJSON files

Reviewer 2: I think we can also mention `minio` and post the `1.5.0` release we can mention Azure as well. 
milestone - https://github.com/astronomer/astro-sdk/milestone/27

Reviewer 2: Also, do you think we should also mention the Databases we support here?

Reviewer 4: We also support loading files from FTP and SFTP locations

Reviewer 3: @jwitz, it may be worth adding a link in case the user wants to confirms the latest supported databases, file locations and types:
https://astro-sdk-python.readthedocs.io/en/latest/supported_databases.html
https://astro-sdk-python.readthedocs.io/en/latest/supported_file.html
https://astro-sdk-python.readthedocs.io/en/latest/supported_file.html#supported-file-type

Author: Yeah links to these docs is better than just creating a long list to parse through, good call!

> - `transform`: Applies a SQL select statement to a source table and saves the result to a destination table. This function allows you to transform your data with a SQL query. It uses a `SELECT` statement that you define to automatically store your results in a new table. By default, the `output_table` is given a unique name each time the DAG runs, but you can overwrite this behavior by defining a specific `output_table` in your function. You can then pass the results of the `transform` downstream to the next task as if it were a native Python object.

Reviewer 3: @jwitz, how extensive do we want the list of SDK functions to be here? Are we happy just having a subset?

Author: I think this doc should still serve as an introduction to the SDK rather than a comprehensive doc, so having only the main functions here is fine with me. If there's an essential/ cool function that you think is missing, I'm happy to add that!

## Installation

Using the Astro Python SDK requires configuring a few things in your Airflow project.

1. Install the Astro Python SDK package in your Airflow environment. If you're using the Astro CLI, add the following to the `requirements.txt` file of your Astro project:
Install the Astro Python SDK package in your Airflow environment. If you're using the Astro CLI, add the following to the `requirements.txt` file of your Astro project:

    ```
    astro-sdk-python
    ```
    
2. If you're using the Astro CLI locally, add the following variable to the `.env` file of your Astro project:

    ```text
    export AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True
    ```

>     Setting `AIRFLOW__CORE__ENABLE_XCOM_PICKLING` to `True` is not recommended in production environments (See [here](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#enable-xcom-pickling)) and should be avoided. Instead, Astronomer recommends using `AIRFLOW__CORE__XCOM_BACKEND`, `AIRFLOW__ASTRO_SDK__XCOM_STORAGE_CONN_ID`, and `AIRFLOW__ASTRO_SDK__XCOM_STORAGE_URL` to allow XCOM data to be saved to a S3 or GCS location. For example:

Author: Do we still need this comment if it's fine to not use pickling? Should we link out to the SDK docs about XCom configuration?

Reviewer 5: Is it still the case that you _have_ to set up an XCom backend if you don't use pickling? If so we should be explicit about that, and in either case I think we should definitely link out to the SDK docs.

Reviewer 2: +1 for linking to SDK docs.

## Installation

1. Install the Astro Python SDK package in your Airflow environment. If you're using the Astro CLI, add the following to the `requirements.txt` file of your Astro project:

    ```
    astro-sdk-python
    ```

> 2. Set the following environment variables. If you're using the Astro CLI, add these environment variables to the `.env` file of your Astro project:

Reviewer 3: It would be nice to explain why users need to set this and give alternatives if:
- they are not using the cloud or
- if they are using GCS

## Installation

1. Install the Astro Python SDK package in your Airflow environment. If you're using the Astro CLI, add the following to the `requirements.txt` file of your Astro project:

    ```
    astro-sdk-python
    ```

2. If you're using Airflow 2.4 or earlier, set the following environment variable to use a required custom XCom backend. If you're using the Astro CLI, add this environment variable to the `.env` file of your Astro project
    
    ```text
    AIRFLOW__CORE__XCOM_BACKEND='astro.custom_backend.astro_custom_backend.AstroCustomXcomBackend'
    ```

> 3. Optional. Create an [Airflow connection](connections.md) to the database where you want to store the temporary tables created by the Astro SDK. After you have successfully tested your connection, set the following environment variables to configure your database as an Astro SDK storage backend. If you're using the Astro CLI, add these environment variables to the `.env` file of your Astro project:

Author: @tatiana @utkarsharma2 Will the SDK work without an external XCom storage destination? Can you just use the defaults for small amounts of data?
